File: idb_companion/Utility/FBStorageUtils.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBStorageUtils.h"

#import <FBControlCore/FBControlCore.h>

#import "FBIDBError.h"

@implementation FBStorageUtils

#pragma mark Finding Files

+ (NSDictionary<NSString *, NSSet<NSURL *> *> *)bucketFilesWithExtensions:(NSSet<NSString *> *)extensions inDirectory:(NSURL *)directory error:(NSError **)error
{
  NSMutableDictionary<NSString *, NSMutableSet<NSURL *> *> *files = NSMutableDictionary.dictionary;
  for (NSString *extension in extensions) {
    files[extension] = NSMutableSet.set;
  }

  NSArray<NSURL *> *contents = [NSFileManager.defaultManager
    contentsOfDirectoryAtURL:directory
    includingPropertiesForKeys:nil
    options:NSDirectoryEnumerationSkipsSubdirectoryDescendants
    error:error];

  if (!contents) {
    return nil;
  }
  for (NSURL *file in contents) {
    NSString *extension = file.pathExtension;
    if (![extensions containsObject:extension]) {
      continue;
    }
    NSMutableSet<NSURL *> *bucket = files[extension];
    [bucket addObject:file];
  }

  return files;
}

+ (NSURL *)findFileWithExtension:(NSString *)extension atURL:(NSURL *)url error:(NSError **)error
{
  NSSet<NSURL *> *files = [self findFilesWithExtension:extension atURL:url error:error];
  if (!files) {
    return nil;
  }
  if (files.count != 1) {
    return [[FBIDBError
      describeFormat:@"%lu files with extension .%@ in %@", (unsigned long)files.count, extension, url]
      fail:error];
  }

  return files.anyObject;
}

+ (NSSet<NSURL *> *)findFilesWithExtension:(NSString *)extension atURL:(NSURL *)url error:(NSError **)error
{
  return [self bucketFilesWithExtensions:[NSSet setWithObject:extension] inDirectory:url error:error][extension];
}

+ (NSURL *)findUniqueFileInDirectory:(NSURL *)directory error:(NSError **)error
{
  NSArray<NSURL *> *filesInDirectory = [self filesInDirectory:directory error:error];
  if (!filesInDirectory) {
    return nil;
  }
  if (filesInDirectory.count != 1) {
    return [[FBIDBError
      describeFormat:@"Expected one top level file, found %lu: %@", filesInDirectory.count, [FBCollectionInformation oneLineDescriptionFromArray:filesInDirectory]]
      fail:error];
  }
  return filesInDirectory[0];
}

+ (NSArray<NSURL *> *)filesInDirectory:(NSURL *)directory error:(NSError **)error
{
  NSError *innerError;
  NSArray<NSURL *> *filesInDirectory = [NSFileManager.defaultManager contentsOfDirectoryAtURL:directory includingPropertiesForKeys:@[NSURLIsDirectoryKey] options:0 error:&innerError];
  if (filesInDirectory == nil) {
    return [[[FBIDBError
      describeFormat:@"Failed to list files in directory %@", directory]
      causedBy:innerError]
      fail:error];
  }
  return filesInDirectory;
}

+ (FBBundleDescriptor *)bundleInDirectory:(NSURL *)directory error:(NSError **)error
{
  NSURL *uniqueFile = [self findUniqueFileInDirectory:directory error:error];
  if (!uniqueFile) {
    return nil;
  }
  return [FBBundleDescriptor bundleFromPath:uniqueFile.path error:error];
}

@end

File: FBControlCore/Utility/FBStorageUtils.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation

public class FBStorageUtils {

  // MARK: Finding Files

  public class func bucketFiles(withExtensions extensions: Set<String>, inDirectory directory: URL) throws -> [String: Set<URL>] {
    var files: [String: Set<URL>] = [:]
    for ext in extensions {
      files[ext] = Set()
    }

    let contents = try FileManager.default.contentsOfDirectory(
      at: directory,
      includingPropertiesForKeys: nil,
      options: .skipsSubdirectoryDescendants
    )

    for file in contents {
      let ext = file.pathExtension
      if extensions.contains(ext) {
        files[ext]?.insert(file)
      }
    }

    return files
  }

  public class func findFile(withExtension ext: String, at url: URL) throws -> URL {
    let files = try findFiles(withExtension: ext, at: url)
    if files.count != 1 {
      throw FBControlCoreError.describe("\(files.count) files with extension .\(ext) in \(url)").build()
    }
    return files.first!
  }

  public class func findFiles(withExtension ext: String, at url: URL) throws -> Set<URL> {
    let buckets = try bucketFiles(withExtensions: Set([ext]), inDirectory: url)
    return buckets[ext] ?? Set()
  }

  public class func findUniqueFile(inDirectory directory: URL) throws -> URL {
    let filesInDirectory = try files(inDirectory: directory)
    if filesInDirectory.count != 1 {
      throw FBControlCoreError.describe("Expected one top level file, found \(filesInDirectory.count): \(FBCollectionInformation.oneLineDescription(from: filesInDirectory))").build()
    }
    return filesInDirectory[0]
  }

  public class func files(inDirectory directory: URL) throws -> [URL] {
    do {
      return try FileManager.default.contentsOfDirectory(at: directory, includingPropertiesForKeys: [.isDirectoryKey], options: [])
    } catch {
      throw FBControlCoreError.describe("Failed to list files in directory \(directory)").caused(by: error as NSError).build()
    }
  }

  public class func bundle(inDirectory directory: URL) throws -> FBBundleDescriptor {
    let uniqueFile = try findUniqueFile(inDirectory: directory)
    return try FBBundleDescriptor.bundle(fromPath: uniqueFile.path)
  }
}
